### PR TITLE
enable/disable user

### PR DESCRIPTION
**Pull request by dvoet**
https://broadinstitute.atlassian.net/browse/DSDEEPB-1386
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: Verify swagger UI on dev server still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala [hunk 1]
```diff
@@ -50,13 +51,37 @@ object UserAuth {
 }
 
 object UserAuthJsonSupport extends JsonSupport {
-  implicit val RawlsUserEmailFormat = jsonFormat1(RawlsUserEmail)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala:53]: these format changes make these "value" objects look like strings

```
{ "userSubjectId": "123", "userEmail": "foo@bar.com" }
```

instead of

```
{ "userSubjectId": {"value": "123"}, "userEmail": {"value": "foo@bar.com"} }
```

File: src/main/scala/org/broadinstitute/dsde/rawls/model/UserModel.scala [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
 import org.joda.time.DateTime
+import UserAuthJsonSupport._
+import spray.httpx.SprayJsonSupport._
 
 /**
  * Created by dvoet on 10/27/15.
  */
 case class UserRefreshToken(refreshToken: String)
 case class UserRefreshTokenDate(refreshTokenUpdatedDate: DateTime)
+case class UserStatus(userInfo: RawlsUser, enabled: Map[String, Boolean])
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/UserModel.scala]: Add a comment explaining what the `enabled` keys are

File: src/main/resources/swagger/rawls.yaml [hunk 3]
```diff
@@ -1920,6 +1920,98 @@ paths:
...
+         '500':
+           description: Rawls Internal Error
+           schema:
+             $ref: '#/definitions/ErrorReport'
+       description: "creates a user in the Rawls system"
+       tags:
+         - users
+       summary: Create user
+       operationId: create
+       parameters:
```
Comment (jmthibault79) [inline src/main/resources/swagger/rawls.yaml]: Add text for the 403/Forbidden case

File: src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala:53]: :+1: I like it

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 4]
```diff
@@ -169,7 +168,7 @@ class HttpGoogleServicesDAO(
         blocking { deleter.execute }
       }) recover {
         //Google returns 409 Conflict if the bucket isn't empty.
-        case gjre: GoogleJsonResponseException if gjre.getStatusCode == 409 =>
+        case gjre: HttpResponseException if gjre.getStatusCode == 409 =>
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala]: variable name no longer makes sense

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 5]
```diff
@@ -475,7 +494,7 @@ class HttpGoogleServicesDAO(
   def toProxyFromUser(subjectId: RawlsUserSubjectId) = s"PROXY_${subjectId.value}@${appsDomain}"
   def toUserFromProxy(proxy: String) = getGroupDirectory.groups().get(proxy).execute().getName
 
-  def adminGroupName = s"${groupsPrefix}-ADMIN@${appsDomain}"
+  def adminGroupName = s"${groupsPrefix}-ADMINS@${appsDomain}"
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:484]: why this change

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 6]
```diff
@@ -21,39 +24,114 @@ object UserService {
...
   case class SetRefreshToken(token: UserRefreshToken) extends UserServiceMessage
   case object GetRefreshTokenDate extends UserServiceMessage
   case object CreateUser extends UserServiceMessage
+  case class GetUserStatus(userRef: RawlsUserRef) extends UserServiceMessage
+  case class EnableUser(userRef: RawlsUserRef) extends UserServiceMessage
+  case class DisableUser(userRef: RawlsUserRef) extends UserServiceMessage
 }
 
-class UserService(userInfo: UserInfo, dataSource: DataSource, googleServicesDAO: GoogleServicesDAO, containerDAO: GraphContainerDAO)(implicit executionContext: ExecutionContext) extends Actor {
+class UserService(protected val userInfo: UserInfo, dataSource: DataSource, protected val gcsDAO: GoogleServicesDAO, containerDAO: GraphContainerDAO, userDirectoryDAO: UserDirectoryDAO)(implicit protected val executionContext: ExecutionContext) extends Actor with AdminSupport with FutureSupport {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:39]: Why `protected`?

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 7]
```diff
@@ -21,39 +24,114 @@ object UserService {
...
 
   def createUser(): Future[PerRequestMessage] = {
-    dataSource.inFutureTransaction() { txn =>
-      val user = RawlsUser(userInfo)
-      googleServicesDAO.createProxyGroup(user) map
-        { _ => containerDAO.authDAO.createUser(user, txn) } map
-        { _ => RequestComplete(StatusCodes.Created) }
+    val user = RawlsUser(userInfo)
+
+    // if there is any error, may leave user in weird state which can be seen with getUserStatus
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:63]: What is the "weird state" recovery plan?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  private def getContext(): InitialDirContext = {
+    val env = new util.Hashtable[String, String]()
+    env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory")
+    env.put(Context.PROVIDER_URL, providerUrl)
+    env.put(Context.SECURITY_PRINCIPAL, user)
+    env.put(Context.SECURITY_CREDENTIALS, password)
+
+    new InitialDirContext(env);
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala]: say no to semicolons

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+    val name = s"mail=${rawlsUser.userSubjectId.value},ou=people,dc=dsde-dev,dc=broadinstitute,dc=org"
+  }
+}
+
+
+/**
+ * this does nothing but throw new OperationNotSupportedException but makes extending classes nice
+ */
+trait BaseDirContext extends DirContext {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala:87]: Is this trait necessary?  I removed it and tests succeeded.

**Comment by jmthibault79** [comment]
Done with review.  Would like opinions from someone with JNDI experience.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala (hunk 5 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:484]: that's what devops called the group

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 6 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:39]: no particular reason, can't be private because it is required by AdminSupport and public did not seem right

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala (hunk 9 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala:87]: it is required, the integration tests will fail

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala (hunk 8 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala]: sorry, this was copied from java

**Comment by dvoet** [comment]
@jmthibault79 if you would like another look, I am done

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 7 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:63]: added comments

**Comment by jmthibault79** [comment]
Looks good to me.
